41-vaccine-temp-monitor/firmware/arduino/src/main.cpp

```diff
@@ -0,0 +1,528 @@
+
+#define CRT_MAX_TEMP_C 30
+#define CRT_SAFE_TEMP_C 25
+#define DEFAULT_SAMPLE_INTERVAL_S 1800
+#define RETRY_S 5
+
+M2M_LM75A lm75a;
+Notecard notecard;
+NotecardEnvVarManager *envVarManager = NULL;
+#ifdef ARDUINO_ARCH_ESP32
+using namespace blues;
+NotecardAuxiliaryWiFi aux_wifi(notecard);
+#endif
+
+volatile int alarm_hysteresis = CRT_SAFE_TEMP_C;
+volatile int alarm_temperature = CRT_MAX_TEMP_C;
+volatile size_t sample_interval_s = DEFAULT_SAMPLE_INTERVAL_S;
+
+void logNoteF(const char *format_, ...)
+{
+    char log[256];
+    va_list args;
+    va_start(args, format_);
+    vsnprintf(log, sizeof(log), format_, args);
+
+    if (J *cmd = notecard.newCommand("hub.log"))
+    {
+        JAddStringToObject(cmd, "text", log);
+        if (!notecard.sendRequest(cmd))
+        {
+            JDelete(cmd);
+        }
+    }
+
+    va_end(args);
+}
+
+int acquireGPSLocation(size_t timeout_s_ = 95)
+{
+    int result;
+    const char *gps_mode;
+    size_t gps_time_s;
+
+    // Get Current Configuration
+    if (J *rsp = notecard.requestAndResponse(notecard.newRequest("card.location.mode")))
+    {
+        gps_mode = JGetString(rsp, "mode");
+        NoteDeleteResponse(rsp);
+    }
+    else
+    {
+        result = -1; // Unable to communicate with Notecard
+        return result;
+    }
+
+    // Halt periodic tracking, before active sync is performed
+    if (J *cmd = notecard.newCommand("card.location.mode"))
+    {
+        JAddStringToObject(cmd, "mode", "off");
+        notecard.sendRequest(cmd);
+    }
+
+    // Gather timestamp of previous location information
+    if (J *rsp = notecard.requestAndResponse(notecard.newRequest("card.location")))
+    {
+        gps_time_s = JGetInt(rsp, "time");
+        logNoteF("INFO: Previous GPS time (seconds): %u", gps_time_s);
+        NoteDeleteResponse(rsp);
+    }
+    else
+    {
+        result = -1; // Unable to communicate with Notecard
+        return result;
+    }
+
+    // Activate GPS
+    if (J *cmd = notecard.newCommand("card.location.mode"))
+    {
+        JAddStringToObject(cmd, "mode", "continuous");
+        notecard.sendRequest(cmd);
+    }
+
+    // Block while resolving GPS location
+    for (const size_t start_ms = ::millis();;)
+    {
+        if (timeout_s_)
+        {
+            if (::millis() >= (start_ms + (timeout_s_ * 1000)))
+            {
+                logNoteF("WARNING: GPS user timeout expired!");
+                result = -3; // User timeout has expired
+                break;
+            }
+        }
+
+        // Check if GPS has acquired location information
+        if (J *rsp = notecard.requestAndResponse(notecard.newRequest("card.location")))
+        {
+            const size_t current_gps_time_s = JGetInt(rsp, "time");
+            if (current_gps_time_s != gps_time_s)
+            {
+                logNoteF("INFO: Current GPS time (seconds): %u", current_gps_time_s);
+                result = 0; // GPS has fixed on position
+                break;

```

Comment:
This leaks rsp, I think. Same below with the other break.

Reply:
I was demonstrating how you can leak memory and it just automatically gets cleaned up when the Notecard controls the power supply of the host. :rofl: